### PR TITLE
Aditya - Fixed Bell Notification for Hours Due in 48 Hours

### PR DESCRIPTION
**Pull request by adityasure9**
# Description:
This pull request resolves the bell notification issue for hours due in 48 hours and includes several new features and improvements. 

image

## Main changes explained:
- Created a new BellNotification.jsx component to handle the logic for displaying notifications when a user has less than 48 hours remaining to meet their weekly committed hours.
- The notification triggers when the user's logged effort is less than 50% and fewer than 48 hours remain in the week.

## How to test:
1. Check into the current branch.
2. Run npm install.
3. Clear Site Data/Cache: This will ensure no old notifications interfere.
4. Log In: Log in as any user.
5. Check the Notification Bell: Navigate to the Dashboard and verify the notification bell functionality.
6. Test Time Manipulation: Use the "Add Intangible Hours" function in the Timelog to manipulate the logged time.
7. Trigger Notification: Verify that the bell notification triggers when less than 50% of hours are logged and only 48 hours remain.
8. Test Reset: Confirm the notification resets after it is viewed or when the week ends.
9. Test Dark Mode: Ensure the notification popup works properly in both light and dark modes.

## Screenshots or videos of changes:

https://github.com/user-attachments/assets/64a1d65e-1c16-4459-9a0c-9c993bdf2505

Add Intangible Time Entry Button:

image

Press the Button and add/edit/delete Time:

image

## Note:
1. Modify Time Left: To test different time thresholds for notifications, modify the hoursLeft value in the BellNotification.jsx component.
image
2. Simulate Time Logs: Use the "Add Intangible Hours" feature to simulate different time log scenarios.
3. Clear Cookies for Retesting: The notification triggers only once per week. Clear cookies or localStorage to reset the notification for retesting.

### CODE REVIEW

**Comment by ShreyaLaheri** [comment]
Hi, everything works well. I have tested for all the above mentioned conditions. And the notification disappears when read, as it should. Great work
image
image
image

**Review by CodewithKoushica — APPROVED** [review]
The new BellNotification.jsx component functions well, triggering notifications as expected when logged hours are under 50% with 48 hours remaining. The reset after viewing or week’s end also works seamlessly.

Feedback:

- The code is clean and modular, with well-placed comments that enhance readability and future maintenance.
- Good use of React best practices, including hooks and prop validation.
- Dark mode and cross-browser compatibility were tested with no issues.

Screenshot 2024-10-21 at 11 03 29 AM
Screenshot 2024-10-21 at 11 03 39 AM

**Review by bhavya17prakash — APPROVED** [review]
The functionality is working as intended.
PR 2805 (2)
PR 2805

**Review by nikhilgiri0226 — APPROVED** [review]
Hi Aditya,

I have tested this PR and did some modification sin the code BellNotification.jsx as you mentioned in order to test the functionality in my local.
image
The notifications icon shows the notifications when the mentioned time is left.
image
The notification resets after we view the notification.
image
image
Dark mode and cross-browser compatibility were tested with no issues.

**Review by Rui-Liu-github — APPROVED** [review]
HI, I reviwed an PR #2805 by pulling the latest version branch from Aditya. I verified that the bell notification triggers when less than 50% of hours are logged and only 48 hours remain. The notification would be reset after being viewed. So everything works as mentioned . A good job. 

image

image

image

**Review by nikhilpittala16 — APPROVED** [review]
I have tested the PR and everything works as expected.

https://github.com/user-attachments/assets/4a1e85f3-36f3-4251-a6c0-0ac9430d9a38

File: src/components/Header/BellNotification.jsx
Comment (sheetalmangate) [inline src/components/Header/BellNotification.jsx:34]: new Date(currentTime)  is returning same value as new Date(). what is the use of passing currentTime?

**Comment by sheetalmangate** [comment]
Code looks good overall. However , I have a concern regarding storage of userId in localStorage.   this could expose sensetive application data. If localStorage is stolen will it harm user privacy or app security?

**Review by zhiminV — DISMISSED** [review]
I have checked the notification bell work, and I can also add/edit/delete Intangible Time. And also work in dark mode. 
Screenshot 2024-10-25 at 11 02 47 PM
Screenshot 2024-10-25 at 11 04 05 PM
Screenshot 2024-10-25 at 11 04 39 PM
Screenshot 2024-10-25 at 11 05 00 PM
Screenshot 2024-10-25 at 11 05 14 PM

**Review by varun-elango — APPROVED** [review]
The code changes look good and the functionality is working as expected. The notification is triggered only when less than 50% of the hours are logged and only 48 hours remain in the week. Otherwise, it is not triggered. The notification goes away after viewing. The notification only triggers once per week. Dark mode and cross-browser compatibility works with no issues.
PR2805_1
PR2805_2
PR2805_4
PR2805_5
PR2805_6

**Comment by Anoushka21** [comment]
The code works as required.
I tested the code in light mode initially and it showed the notification and the notification disappears once seen. It also remains disappeared in dark mode and vice-versa. Good job!

https://github.com/user-attachments/assets/d0339421-a3f6-47fc-acf8-4a130a253dbb

**Review by Anoushka21 — DISMISSED** [review]

**Review by RachanaZha — DISMISSED** [review]
Have checked the changes. It works perfectly fine and after adding the time  more then 50%(i added using intangible time entry), the notification disappears. Have checked in both light and dark mode, working perfectly fine!
zha_rachana_PR#2805
zha_rachana_PR#2805_ss2
zha_rachana_PR#2805_ss3

**Comment by VMD281** [comment]
Hi, I reviewed this PR and the bell feature works perfectly. After adding the time less than 50%
PR#2805__3
PR#2805_1
PR#2805_2
, the notification disappears. Have checked in both light and dark mode, working perfectly fine!

**Review by Rupa2400 — DISMISSED** [review]
I have verified the Notification Bell functionality by navigating to the Dashboard. The bell notification successfully triggers when less than 50% of hours are logged, and only 48 hours remain. Additionally, the notification resets appropriately after being viewed or at the end of the week. I also confirmed that the notification popup operates correctly in both light and dark modes. Great work!

PR_2805

PR_2805_2

PR_2805_3

**Review by anji-annavarapu — DISMISSED** [review]

**Review by hqzhu0913 — APPROVED** [review]
The bell notification will appear as 50% less than assigned hour and deadline due in 48 hours. It functions well in both light and dark mode.
Screenshot 2024-10-25 at 3 46 58 PM
Screenshot 2024-10-25 at 3 47 28 PM

File: src/components/Header/BellNotification.jsx
Comment (adityasure9) [reply src/components/Header/BellNotification.jsx:34]: Hey Sheetal, Good observation! You’re absolutely right; passing currentTime in that line is redundant since it's already a Date object. We could just use new Date() directly, which keeps the code a bit cleaner without changing any functionality. Thanks for catching that!

**Review by AshritaCherlapally — APPROVED** [review]
I have reviewed the new BellNotification.jsx component, which effectively handles the logic for displaying notifications for users with less than 48 hours remaining to meet their weekly committed hours. It is working as expected, also the Intangible add time entry is also working fine, I have tested by logging in a task, tried to edit and deleted the log. 

https://github.com/user-attachments/assets/c868999d-97c5-4ecd-96ce-3b6c499310b1

**Review by audreydieuanh — DISMISSED** [review]
The bell notification works correctly and disappears after being seen as expected.
Screenshot 2024-11-02 at 9 12 23 PM

**Comment by KurtisIvey** [comment]
Hi, I noticed there was a merge conflict on this. You've got a lot of approvals on this pull request and we'd love to get this moved to final review and merged once that merge conflict is resolved. Thank you

**Comment by CodewithKoushica** [comment]
Updated the file for few changes and created a new [PR #2961](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2961)

**Pull request closed without merging**
